subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junit/RerunPreviousFailedTestIntegrationTest.groovy

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.testing.junit
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.DefaultTestExecutionResult
+import org.hamcrest.Matchers
+import spock.lang.Unroll
+
+class RerunPreviousFailedTestIntegrationTest extends AbstractIntegrationSpec {
+    static final String INDEX_OF_TEST_TO_FAIL = "index.of.test.to.fail"
+
+    def setup() {
+        buildFile << """
+            apply plugin: 'java'
+
+            ${mavenCentralRepository()}
+
+
+            dependencies {
+                testCompile 'junit:junit:4.12'
+            }
+
+            tasks.withType(Test) {
+                maxParallelForks = 1 
+            }
+        """
+
+        (1..3).each {
+            file("src/test/java/ConditionalFailingTest_${it}.java") << """
+                import org.junit.Test;
+                public class ConditionalFailingTest_${it} {
+                    @Test
+                    public void failedTest() {
+                        if("${it}".equals(System.getProperty("${INDEX_OF_TEST_TO_FAIL}"))) {
+                            throw new RuntimeException();
+                        }
+                    }
+                }
+            """.stripIndent()
+        }
+    }
+
+    def letTestFail(def index) {
+        buildFile << """
+        test {
+            systemProperty('${INDEX_OF_TEST_TO_FAIL}', '${index}')
+        }
+        """
+    }
+
+    @Unroll
+    def 'subsequent fail fast execution runs failed test first'() {
```

Comment:
⭕️ I think this test is testing too many things at once. It tests --fail-fast, but also a normal test. Should probably be two test cases.

Reply:
I don't think it's testing `--fail-fast`. This test **depends on** the right behavior of `--fail-fast` so I don't say it's **testing** `--fail-fast`.

Reply:
Sorry, but that's splitting hairs. My point is that this test tests two things:

- failed tests are run first with `--fail-fast`
- failed tests are run first without `--fail-fast`

These are two different test cases that belong in different methods. I thought we were missing test coverage when I only looked at the test names.

Reply:
Sorry I should have made the test case name clearer. Now it's `subsequent execution runs failed test first`.

Reply:
This is still testing two scenarios in one test:

- failed tests are run first with --fail-fast
- failed tests are run first without --fail-fast

These should be two test cases to make the tests easier to follow and to make it clear that all important scenarios are covered by just reading the test names, not reading the full test implementation.

Reply:
Sorry maybe I didn't make myself clear here. We want subsequent test to stop immediately when first failed test case occurs, so that we can analyze if this failed test case is exactly the one which failed in previous test. That's why I used `--fail-fast` here - it's used to guarantee the subsequent test stops on first failure. To exclude the impact of original test case order, a `@Unroll` test is used here, i.e. what if the failed test case happens to be the first one to run in subsequent test?

Reply:
I see, so you are using fail-fast so you can deduce the order by looking at the state of the test? That's pretty hard to understand by just looking at the code to be honest. And it could break if we change the reporting of fail-fast. 

Using the BlockingHttpServer would need a bit more code, but might make this condition clearer. I.e. you could block the other tests until the failed one has executed.